public/app/features/alerting/unified/hooks/useFilteredRules.ts

```diff
@@ -0,0 +1,60 @@
+import { useMemo } from 'react';
+
+import { CombinedRuleGroup, CombinedRuleNamespace } from 'app/types/unified-alerting';
+import { isCloudRulesSource } from '../utils/datasource';
+import { isAlertingRule } from '../utils/rules';
+import { getFiltersFromUrlParams } from '../utils/misc';
+import { useQueryParams } from 'app/core/hooks/useQueryParams';
+
+export const useFilteredRules = (namespaces: CombinedRuleNamespace[]) => {
+  const [queryParams] = useQueryParams();
+  const filters = getFiltersFromUrlParams(queryParams);
+
+  return useMemo(() => {
+    const filteredNamespaces = namespaces
```

Comment:
This is...quite the mouthful. Could you break it out into more functions and perhaps add comments where it makes sense?